core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -0,0 +1,184 @@
+      Utils.classForName(parameters(SINK_CALLBACK)).newInstance()
+    } catch {
+      case _: NoSuchElementException => new DefaultSnappySinkCallback()
+    }
+
+    SnappyStoreSink(sqlContext.asInstanceOf[SnappyContext].snappySession, parameters,
+      cc.asInstanceOf[SnappySinkCallback])
+  }
+
+  private def createSinkStateTableIfNotExist(sqlContext: SQLContext) = {
+    sqlContext.asInstanceOf[SnappyContext].snappySession.sql(s"create table if not exists" +
+      s" $SINK_STATE_TABLE (" +
+      " stream_query_id varchar(200)," +
+      " batch_id long, " +
+      " PRIMARY KEY (stream_query_id)) using row options(DISKSTORE 'GFXD-DD-DISKSTORE')")
+  }
+
+  @Override
+  def shortName(): String = SNAPPY_SINK_NAME
+
+}
+
+case class SnappyStoreSink(snappySession: SnappySession,
+    parameters: Map[String, String], sinkCallback: SnappySinkCallback) extends Sink {
+
+  override def addBatch(batchId: Long, data: Dataset[Row]): Unit = {
+    val streamQueryId = parameters(STREAM_QUERY_ID).toUpperCase
+
+    val updated = snappySession.sql(s"update $SINK_STATE_TABLE " +
+      s"set batch_id=$batchId where stream_query_id='$streamQueryId' and batch_id != $batchId")
+      .collect()(0).getAs("count").asInstanceOf[Long]
+
+    // TODO: use JDBC connection here
+    var posDup = false
+
+    if (updated == 0) {
+      try {
+        snappySession.insert(SINK_STATE_TABLE, Row(streamQueryId, batchId))
+        posDup = false
+      }
+      catch {
+        case e: SQLException if e.getSQLState.equals("23505") => posDup = true
+        case exception: Throwable => throw exception
+      }
+    }
+
+    sinkCallback.process(snappySession, parameters, batchId, convert(data), posDup)
+  }
+
+  /**
+    * This conversion is necessary as Sink
+    * documentation disallows an operation on incoming dataframe.
+    * Otherwise it will break incremental planning of streaming dataframes.
+    * See http://apache-spark-developers-list.1001551.n3.nabble.com/
+    * Structured-Streaming-Sink-in-2-0-collect-foreach-restrictions-added-in-
+    * SPARK-16020-td18118.html
+    * for a detailed discussion.
+    */
+  def convert(ds: DataFrame): DataFrame = {
+    snappySession.internalCreateDataFrame(
+      ds.queryExecution.toRdd,
+      StructType(ds.schema.fields))
+  }
+}
+
+object DefaultSnappySinkCallback {
+  private val log = Logger.getLogger(classOf[DefaultSnappySinkCallback].getName)
+}
+
+import org.apache.spark.sql.snappy._
+
+class DefaultSnappySinkCallback extends SnappySinkCallback {
+  def process(snappySession: SnappySession, parameters: Map[String, String],
+              batchId: Long, df: Dataset[Row], posDup: Boolean) {
+
+    df.cache()
+    val snappyTable = parameters(TABLE_NAME).toUpperCase
+    DefaultSnappySinkCallback.log.debug(s"Processing for $snappyTable and batchId $batchId")
+
+    val tableName = parameters(TABLE_NAME).toUpperCase
+    val keyColumnsDefined = !snappySession.sessionCatalog.getKeyColumns(tableName).head(1).isEmpty
```

Comment:
This creates a newDataFrame and will result in yet another job being created which is an unnecessary overhead. Refactor the sessionCatalog.getKeyColumns to return the Seq[Column] (say getKeyColumnsSeq) and use that here.